[KOGITO-8353] Introduce Camel Routes integration guide (#269)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/integrations/camel-routes-integration.adoc

> = Integrating with Camel Routes

Reviewer: ```suggestion
= Integrating with Camel routes
```

> {context} can integrate with link:{camel_url}[Apache Camel Routes] by adding the Kogito Quarkus Camel Add-on to your project. It enables the workflow engine to identify and call Camel Routes declared in YAML or XML in the same workflow project context.

Reviewer: ```suggestion
{context} can integrate with link:{camel_url}[Apache Camel Routes] by adding the Kogito Quarkus Camel Add-on to your project. It enables the workflow engine to identify and call Camel routes declared in YAML or XML in the same workflow project context.
```

[[proc-enable-quarkus-camel]]
== Enabling Quarkus Camel in {context}

You can enable Quarkus Camel in your project.

.Prerequisites
* A workflow application is created. 
+
For more information about creating a workflow, see xref:getting-started/create-your-first-workflow-service.adoc[Creating your first workflow service].

.Procedure

> . To add the Quarkus Camel to your workflow application, add the `org.kie.kogito:kogito-addons-quarkus-camel` dependency to the `pom.xml` file of your project:

Reviewer: ```suggestion
. To add the Quarkus Camel to your workflow application, add the `org.kie.kogito:kogito-addons-quarkus-camel` dependency to the `pom.xml` file of your project as follows:
```

+
--

> .Dependency to be added to the `pom.xml` file to enable metrics

Author: This is the same writing pattern as other guides. We can review in a new PR for every other one. I'll do it sometime this week after we merge this.

> .Dependency to be added to the `pom.xml` file to enable metrics

Reviewer 2: sure ! I was suggesting as some other guides have the different possibilities

[source,xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-camel</artifactId>
</dependency>
----
--

[[con-creating-camel-routes]]

> == Creating Camel Routes in {context}

Reviewer: ```suggestion
== Creating Camel routes in {context}
```

> You can add YAML or XML Camel Routes in your workflow project.

Reviewer: ```suggestion
You can add YAML or XML Camel routes to your workflow project.
```

.Procedure
. Create a YAML or XML Camel Routes using your IDE or the link:{kaoto_url}[Kaoto VSCode Editor] and place them in the `src/main/resources/routes` directory.

> . The route `from` endpoint *must* be a `direct` component. That's the endpoint producer expected by the workflow engine.

Reviewer: ```suggestion
. The route `from` endpoint must be a `direct` component. That's the endpoint producer expected by the workflow engine.
```

. The route response must be in a valid format that the workflow context can understand:
+
include::../../pages/_common-content/camel-valid-responses.adoc[]
+

> The response will be merged to the workflow state context. If it's an array or a complex object, the response will be added to the special attribute `response`.

Reviewer: ```suggestion
The response will be merged into the workflow state context. If it is an array or a complex object, the response will be added to the special attribute `response`.
```

> .Example of a Camel Route with the `direct` endpoint returning a valid JSON string representation

Reviewer: ```suggestion
.Example of a Camel route with the `direct` endpoint returning a valid JSON string representation
```

[source,yaml]
----
- from:
    uri: direct:logRouteReplaceHeader
    steps:
      - setBody:
          simple: '{ "id": "${header.WorkflowID}", "arg1": { "arg2": "value1" } }'
      - log:
          message: We received the ${body}

> ----

Reviewer 2: Maybe explain a bit what happens here in that route to make sure the user does not spend too much time understanding the context here ...

[[con-importing-camel-routes]]

> == Defining and Referencing Camel functions in the Workflow DSL

Reviewer: ```suggestion
== Defining and referencing Camel functions in the Workflow DSL
```

You can define and reference your Camel functions in the workflow definition.

.Prerequisites

> * You have created Camel Routes in the workflow Maven project.

Reviewer: ```suggestion
* You have created Camel routes in the workflow Maven project.
```

.Procedure

> . In the `functions` definition section of your workflow DSL, declare the Camel Route like exemplified below:

Reviewer: ```suggestion
. In the `functions` definition section of your workflow DSL, declare the Camel route as exemplified below:
```

+
.Example of a Camel Route function definition
[source,json]
----
{
  "functions": [
    {
      "name": "logRoute",
      "type": "custom",
      "operation": "camel:direct:logRouteReplaceHeader"
    }
  ]
}
----
+

> The operation description must have the prefix `camel:direct:`, indicating that you want to produce a message to this route via the link:{camel_extensions_url}/direct.html[Camel Direct Component]. Direct is the only component supported by {context} at the moment.

Reviewer 2: We should have a page with the Camel supported components instead of writing it into the text, which we may forget about ...

Author: We won't support anything else. If so, we can then update the guide with a supported components.

Reviewer 2: ok

+
The `operation` suffix contains the name of the route endpoint. In the case of this example, `logRouteReplaceHeader`.

[WARNING]
====

> The Camel Route defined in the workflow must be available in your project during runtime, otherwise an `IllegalArgumentException` will thrown.

Reviewer: ```suggestion
The Camel route defined in the workflow must be available in your project during runtime, otherwise, an `IllegalArgumentException` will be thrown.
```

====

> . To use the Camel function definition in a workflow action, you can simply reference to it as you normally would with any other {context} function definitions. For example:

Reviewer: ```suggestion
. To use the Camel function definition in a workflow action, you can simply reference it as you normally would with any other {context} function definitions. For example:
```

> . To use the Camel function definition in a workflow action, you can simply reference it as you normally would with any other {context} function definitions. For example:

Reviewer 2: numbering here is wrong in the demo from netlify

![image](https://user-images.githubusercontent.com/17157711/208699701-5c0c893c-5fbc-4c58-a594-84e69627f40c.png)

It should be `2.`

+
.Example of a workflow state action referencing
[source,json]
----
{
  "states": [
    {
      "name": "sendToLog",
      "type": "operation",
      "actions": [
        {
          "functionRef": {
            "refName": "logRoute", <1>
            "arguments": {
              "body": "${ . }", <2>
              "headers": {
                "WorkflowID": "$WORKFLOW.instanceId" <3>
              }
            }
          }
        }
      ],
      "end": true
    }
  ]
}
----
+
<1> The function reference name, as defined in the function attribute.
<2> The body definition. A `jq` expression to be evaluated in runtime resulting in the body payload.
<3> The `headers` definition that must be a key/value pair or a valid `jq` expression evaluated in runtime.

> Once a message is received back from the Camel route, the data is merged to the workflow payload.

Reviewer: ```suggestion
Once a message is received back from the Camel route, the data is merged into the workflow payload.
```

> Once a message is received back from the Camel route, the data is merged into the workflow payload.

Reviewer 2: What is the output of all of this ? The user will want to know ! :)

> == Example Project

Reviewer: ```suggestion
== Example project
```

> There's an link:{kogito_sw_examples_url}/serverless-workflow-camel-routes[example project available on GitHub] using this new feature. You can use it as a reference to have a better understanding of the Camel integration with {context}.

Reviewer: ```suggestion
There is an link:{kogito_sw_examples_url}/serverless-workflow-camel-routes[example project available on GitHub] using this new feature. You can use it as a reference to have a better understanding of the Camel integration with {context}.
```

== Additional resources

> * xref:core/custom-functions-support[Custom functions for your {context} service]

Reviewer 2: link is wrong in the netlify demo... not sure why ...
